kratos/utilities/constraint.h

```diff
@@ -0,0 +1,294 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(CONSTRAINT_H)
+#define CONSTRAINT_H
+// System includes
+#include <vector>
+#include <unordered_map>
+#include <iostream>
+#include <tuple>
+#include <utility>
+#include <assert.h>
+
+// project includes
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include "geometries/geometry.h"
+#include "containers/constraint_data.h"
+#include "containers/variable.h"
+#include "containers/variable_component.h"
+#include "containers/vector_component_adaptor.h"
+
+namespace Kratos
+{
+/** \brief Constraint
+	* A class that implements the interface for different constraints to be applied on a system.
+    */
+
+class ProcessInfo;
+class Element;
+class Condition;
+
+template <class TSparseSpace,
+          class TDenseSpace //= DenseSpace<double>
+          >
+class Constraint
+{
+
+  public:
+    /// Pointer definition of DataValueContainer
+    KRATOS_CLASS_POINTER_DEFINITION(Constraint);
+
+    typedef Dof<double> DofType;
+    typedef Node<3> NodeType;
+    typedef PointerVectorSet<NodeType, IndexedObject> NodesContainerType;
+    typedef std::vector<std::size_t> EquationIdVectorType;
+
+    typedef typename TDenseSpace::MatrixType LocalSystemMatrixType;
+    typedef typename TDenseSpace::VectorType LocalSystemVectorType;
+    typedef typename TSparseSpace::DataType TDataType;
+    typedef typename TSparseSpace::MatrixType TSystemMatrixType;
+    typedef typename TSparseSpace::VectorType TSystemVectorType;
+
+    ///@name Life Cycle
+    ///@{
+
+    /**
+	* Creates a Constraint object
+	*/
+    Constraint(std::string iName = "default", bool iIsActive = true) : mActive(iIsActive), mName(iName)
+    {
+    }
+    /// Destructor.
+    virtual ~Constraint(){};
+
+    ///@}
+
+    ///@name Access
+    ///@{
+
+    /**
+	* Clears the maps contents
+	*/
+    void Clear()
+    {
+        mConstraintEquationContainer.Clear();
+    }
+
+    /**
+	* Get the MasterDOFs vector for this slave
+	* @return MasterDOFs vector for this slave
+	*/
+    virtual const ConstraintEquation &GetConstraintEquation (DofType &SlaveDof)

```

Comment:
This isn't needed by the builder and solver. It creates the dependencies on the other constraint classes and should be moved to another class.

Reply:
@msandre  As you mentioned  this is not necessary for the builder and solver. But this function is necessary for the specific implementations of the constraints to access the specific constraint equations and access the DOFs involved. If we look into multipoint_constraint.h we see this is used.

Reply:
This is related to the comment above. You can decide, just consider that this adds dependencies to builder and solver.